mixer/adapter/kubernetes2/cache.go

```diff
@@ -0,0 +1,266 @@
+	}
+
+	// used to send updates to the logger
+	resourceMutation struct {
+		kind eventType
+		obj  interface{}
+	}
+
+	eventType int
+)
+
+const (
+	addition eventType = iota
+	update
+	deletion
+)
+
+// mutationBufferSize sets the limit on how many mutation events can be
+// outstanding at any moment. 100 is chosen as a reasonable default to start.
+// TODO: make this configurable
+const mutationBufferSize = 100
+
+// errorDelay controls how long the logger waits when encountering an error
+// during logging. This should only ever happen when the underlying cache has
+// not yet synced (at which point we need to wait before doing any further
+// processing).
+// TODO: make this configurable
+const errorDelay = 1 * time.Second
+
+const debugVerbosityLevel = 4
+
+// Responsible for setting up the cacheController, based on the supplied client.
+// It configures the index informer to list/watch pods and send update events
+// to a mutations channel for processing (in this case, logging).
+func newCacheController(clientset *kubernetes.Clientset, refreshDuration time.Duration, env adapter.Env) cacheController {
+	c := &controllerImpl{
+		clientset:     clientset,
+		env:           env,
+		mutationsChan: make(chan resourceMutation, mutationBufferSize),
+		ipPodMapMutex: &sync.RWMutex{},
+		ipPodMap:      make(map[string]string),
+	}
+
+	namespace := "" // todo: address unparam linter issue
+
+	c.pods = cache.NewSharedIndexInformer(
+		&cache.ListWatch{
+			ListFunc: func(opts metav1.ListOptions) (runtime.Object, error) {
+				return clientset.Pods(namespace).List(opts)
+			},
+			WatchFunc: func(opts metav1.ListOptions) (watch.Interface, error) {
+				return clientset.Pods(namespace).Watch(opts)
+			},
+		},
+		&v1.Pod{},
+		refreshDuration,
+		cache.Indexers{},
+	)
+
+	c.pods.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.updateIPPodMap,
+			DeleteFunc: c.deleteFromIPPodMap,
+			UpdateFunc: func(old, cur interface{}) {
+				if !reflect.DeepEqual(old, cur) {
+					c.updateIPPodMap(cur)
+				}
+			},
+		},
+	)
+
+	// debug logging for pod update events
+	if env.Logger().VerbosityLevel(debugVerbosityLevel) {
+		c.pods.AddEventHandler(
+			cache.ResourceEventHandlerFuncs{
+				AddFunc: func(obj interface{}) {
+					c.mutationsChan <- resourceMutation{addition, obj}
+				},
+				DeleteFunc: func(obj interface{}) {
+					c.mutationsChan <- resourceMutation{deletion, obj}
+				},
+				UpdateFunc: func(old, cur interface{}) {
+					if !reflect.DeepEqual(old, cur) {
+						c.mutationsChan <- resourceMutation{update, cur}
+					}
+				},
+			},
+		)
+	}
+
+	return c
+}
+
+// Run starts the logger and the controller for the pod cache.
+func (c *controllerImpl) Run(stop <-chan struct{}) {
+	if c.env.Logger().VerbosityLevel(debugVerbosityLevel) {
+		c.env.ScheduleDaemon(func() {
+			c.runLogger(stop)
+		})
+	}
+	c.env.ScheduleDaemon(func() {
+		c.pods.Run(stop)
+		c.env.Logger().Infof("pod cache started")
+	})
+	<-stop
```

Comment:
If it synchronously wait for stop, I don't think we should schedule `c.pods.Run` in another goroutine, simply invoke here directly.